### PR TITLE
Unify plugin hooks dispatch

### PR DESCRIPTION
**Pull request by dyrock**
1. Added`HttpHookState` class that manages all http hooks. This will be useful for later plugin coordination work.
2. Changed `ProxyClientSession` and `HttpSM` to use `HttpHookState` for hooks dispatch
3. Minor changes to `APIHook`, `APIHooks`, and `FeatureAPIHooks`

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
This is phase one of the [Plugin Coordination project](https://solidwallofcode.github.io/plugin-coordination.en.html). The early phases will provide mostly debugging support.

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

File: proxy/InkAPIInternal.h [hunk 1]
```diff
@@ -213,28 +220,18 @@ template <typename ID, ID N>
...
-  hooks_p = false;
-}
-
-template <typename ID, ID N>
-void
-FeatureAPIHooks<ID, N>::prepend(ID id, INKContInternal *cont)
-{
-  if (likely(is_valid(id))) {
-    hooks_p = true;
-    m_hooks[id].prepend(cont);
```
Comment (shinrich) [inline proxy/InkAPIInternal.h]: I assume the prepend code is just never used anywhere?

File: proxy/ProxyClientSession.cc [hunk 2]
```diff
@@ -124,43 +118,35 @@ ProxyClientSession::state_api_callout(int event, void *data)
...
-            return 0;
-          }
-        }
+      Ptr<ProxyMutex> plugin_mutex;
 
-        this->api_current = this->api_current->next();
-        hook->invoke(eventmap[this->api_hookid], this);
+      if (hook->m_cont->mutex) {
+        plugin_mutex = hook->m_cont->mutex;
+        plugin_lock  = MUTEX_TAKE_TRY_LOCK(hook->m_cont->mutex, mutex->thread_holding);
```
Comment (shinrich) [inline proxy/ProxyClientSession.cc]: While you are in here, you may want to look at replacing MUTEX_TAKE_TRY_LOCK with MUTEX_TRY_LOCK.  I see that MUTEX_TAKE_TRY_LOCK has been deprecated.  But given the structure of this code with the possibility of a null mutex pointer, that may not be a very straightforward replacement.

**Comment by shinrich** [comment]
Overall this cleanup looks good.  I think our current set of autests exercises both session and transaction callbacks, so that would flush out any obvious errors.

**Comment by SolidWallOfCode** [comment]
Otherwise this looks reasonable. I'll leave it up for a bit more for more comment.

File: proxy/ProxyClientSession.cc [hunk 3]
```diff
@@ -124,43 +118,35 @@ ProxyClientSession::state_api_callout(int event, void *data)
...
-          Mutex_unlock(plugin_mutex, this_ethread());
+        if (!lock.is_locked()) {
+          SET_HANDLER(&ProxyClientSession::state_api_callout);
+          mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
+          return 0;
         }
+      }
+
+      cur_hook = nullptr; // mark current callback at dispatched.
+      hook->invoke(eventmap[hook_state.id()], this);
```
Comment (ywkaras) [inline proxy/ProxyClientSession.cc]: The presence of eventmap array is very misleading.  This code only ever handles the SSN_START and SSN_CLOSE hooks.  It's OK if you don't want to remove this as a part of this PR.  I'm hoping to get around to removing it eventually.

**Comment by ywkaras** [comment]
It seems there are no new gold tests in this PR.  Are there existing gold tests that will sufficiently exercise this code?

File: proxy/ProxyClientSession.cc (hunk 3 above)
Comment (SolidWallOfCode) [reply proxy/ProxyClientSession.cc]: How is it misleading? It translates from the hook ID to the event code.

Comment (ywkaras) [reply proxy/ProxyClientSession.cc]: As I said it implies that this code path handles a large number of events rather than just SSN_START and SSN_CLOSE.

**Comment by SolidWallOfCode** [comment]
Large failures on AuTest. Rebase and try again, there was a non-trivial change recently.

**Pull request closed without merging**